File: desktop/src/utils/alertInput.test.js
import { test } from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import { parseTarget, parsePct, parseMinEur, toInput } from './alertInput.js';

// ZWILLING: android/app/src/test/java/com/example/yugiohscanner/AlertInputTest.kt liest dieselbe Fixture.
const FIX = JSON.parse(fs.readFileSync(new URL('../../../docs/fixtures/portfolio/alert-input.json', import.meta.url), 'utf8'));

const check = (fn, cases) => {
  for (const c of cases) assert.deepEqual(fn(c.in), { value: c.value, error: c.error }, JSON.stringify(c.in));
};

test('Zielpreis', () => check(parseTarget, FIX.target));
test('Prozent', () => check(parsePct, FIX.pct));
test('Mindestbetrag', () => check(parseMinEur, FIX.minEur));
test('Anzeige im Feld', () => {
  for (const c of FIX.toInput) assert.equal(toInput(c.in), c.out);
});

File: android/app/src/test/java/com/example/yugiohscanner/AlertInputTest.kt
package com.example.yugiohscanner

import com.example.yugiohscanner.ml.AlertInput
import com.example.yugiohscanner.ml.AlertParse
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Test

/** ZWILLING von desktop/src/utils/alertInput.test.js -- dieselbe Fixture docs/fixtures/portfolio/alert-input.json. */
class AlertInputTest {
    private val fix = JSONObject(Fixtures.text("docs/fixtures/portfolio/alert-input.json"))

    private fun check(cases: JSONArray, fn: (String) -> AlertParse) {
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            val input = c.getString("in")
            val expected = AlertParse(
                if (c.isNull("value")) null else c.getDouble("value"),
                if (c.isNull("error")) null else c.getString("error"),
            )
            assertEquals("\"$input\"", expected, fn(input))
        }
    }

    @Test fun `Zielpreis`() = check(fix.getJSONArray("target")) { AlertInput.parseTarget(it) }
    @Test fun `Prozent`() = check(fix.getJSONArray("pct")) { AlertInput.parsePct(it) }
    @Test fun `Mindestbetrag`() = check(fix.getJSONArray("minEur")) { AlertInput.parseMinEur(it) }

    @Test fun `Anzeige im Feld`() {
        val cases = fix.getJSONArray("toInput")
        for (i in 0 until cases.length()) {
            val c = cases.getJSONObject(i)
            assertEquals(c.getString("out"), AlertInput.toInput(if (c.isNull("in")) null else c.getDouble("in")))
        }
    }
}
